docs/apis/subsystems/group/index.md

```diff
@@ -0,0 +1,140 @@
+
+- Find out the current settings for this instance of the activity
+- Show group controls (for example group selection menus) when appropriate
+- Explore memberships and structures of groups
+- Modify it's own interface to hide/show data accordingly
+
+:::note
+
+Groups are typically only relevant to course features such as Activity modules, and some blocks.
+
+Some other core subsystems also need to be group-aware.
+
+:::
+
+## Group modes
+
+There are three different group modes, these modes allow for restrictions to be put in place for access and visibility.
+
+- No groups - The course or activity has no groups.
+- Separate groups - Teachers and students can normally only see information relevant to that group.
+- Visible groups - Teachers and students are separated into groups but can still see all information.
+
+This is explained in more detail on the [Groups access control](https://docs.moodle.org/dev/Groups_access_control) page.
+
+## File locations
+
+The Groups API is currently defined in [lib/grouplib.php](https://github.com/moodle/moodle/blob/master/lib/grouplib.php). This contains global functions which have the `groups_` prefix, for example: `groups_get_group()`.
+
+## Examples
+
+### How to find and use the "current" group
+
+This is using an example from the module forums.
+
+```php
+// Get the course module id from a post or get request.
+$id = required_param('id', PARAM_INT);
+
+// Get the course module.
+$cm = get_coursemodule_from_id('forum', $id, 0, false, MUST_EXIST)
+
+// Get the current group id.
+$currentgroupid = groups_get_activity_group($cm);
+// Get the current group name from the group id.
+$currentgroupname = groups_get_group_name($currentgroupid);
+
+// Do as you please with your newly obtained group information.
+```
+
+### How to make sure that the current user can see a given item in your activity
+
+The following example:
+
+- fetches the course module record for the specified forum id
+- checks whether it has a group mode specified (separate or visible groups)
+- fetches the list of possible groups for that activity
+- checks whether the forum discussion is in a valid group
+
+For this example we are going to check to see if groups are active and whether the user has access to the discussion.
+
+```php
+// Get the course module and discussion id from a post or get request.
+$id           = required_param('id', PARAM_INT);
+$discussionid = required_param('discussionid', PARAM_INT);
+
+// Get the course module.
+$cm = get_coursemodule_from_id('forum', $id, 0, false, MUST_EXIST);
+
+// Get the group id for this discussion
+$discussiongroup = $DB->get_record('forum_discussions', ['id' => $discussionid], groupid);
+
+// Check access.
+if (groups_get_activity_groupmode($cm)) {
+    $groups = groups_get_activity_allowed_groups($cm);
+} else {
+    $groups = [];
+}
+if (!in_array($discussiongroup->groupid, array_keys($groups))) {
+    print_error('groupnotamember', 'group');
```

Comment:
I found this example a bit confusing. If the forum is in group mode "NOGROUPS" (value zero) it won't let the user access it.

Reply:
If it's in no group mode, then the user shoudl not have access to discussions which have a group b.

Reply:
I think it is a strange example. If it checks the group mode only if the post has a group id it will feel more real.

Reply:
But that would be wrong. If a discussion has a group, and I am not in that group, then I should not be able to access it (except with the accessallgroups capability).

In any case, this is a long standing and existing example. It is 100% correct and was in the codebase until about 12 months ago when the code structure was changed but the same logic is still in place.

Remember that a discussion may have been in a group and now the course reconfigured. It would be wrong to display that discussion to users to whom it was not originally intended

Reply:
OK then. I miss some examples using the group mode constants but this is something we can add later. There is no reason to delay this pull request for this.